### ISSUE TITLE
I want to achieve the same thing as this issue but have more questions https://github.com/ultralytics/ultralytics/issues/1631

### ISSUE THREAD

**Issue by shahbazahmed108** [ultralytics/ultralytics#7448]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

this is the issue that I want to follow 

https://github.com/ultralytics/ultralytics/issues/1631

this was the original author's solution when he mentioned the creator

@glenn-jocher thank you I did it. I copied the best.pt and paste it to main folder with different name and I change the model name as you said and it worked.

yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m.pt** imgsz=640 batch=11 patience=64

yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m_custom_train1.pt** imgsz=640 batch=11 patience=64

Like I said in the title, I am trying to achieve the same thing. but the issue is that when I label the new dataset, I have the same classes as my pre-trained model and new classes. the index of the common classes across old and new datasets .yaml file doesn't match as in my pre-trained model label dataset and new label dataset. does this cause an issue, if yes, how to solve this. Thank you

**Comment by glenn-jocher** [comment]
@shahbazahmed108 hello! It's great to see you're building upon previous solutions. When combining datasets with different class indices, it's important to ensure that the class indices match between your pre-trained model and the new dataset. Mismatched indices can indeed cause issues, as the model won't correctly associate the learned features with the appropriate classes.

To solve this, you'll need to align the class indices in your new dataset's `.yaml` file with those of the pre-trained model's dataset. Make sure that the common classes have the same index numbers, and append the new classes to the end of the list. This way, your model can correctly apply its learned knowledge to the common classes and start learning the new ones during further training.

If you need more detailed guidance on dataset preparation or training, please refer to our documentation on custom dataset training and the relevant sections on configuring your `.yaml` files. Good luck with your training! 🚀

**Comment by ehsanadeli1** [comment]
@glenn-jocher Hi! I have a naive question related to this discussion. I want to train YOLOv8m on custom data where the object is none of the 80 predefined objects. Let's say the license plate is my object. Should my *.yaml file look like this:

# data.yaml
train: ./train/images
val: ./valid/images
# Classes
names:
  0: license plate


Or should it look like this:

# data.yaml
train: ./train/images
val: ./valid/images
# Classes
names:
  80: license plate

Considering that the first 79 classes are reserved (https://docs.ultralytics.com/datasets/detect/coco/#applications)."

**Comment by glenn-jocher** [comment]
@ehsanadeli1 hello! For training YOLOv8m on a custom dataset with a new object class such as 'license plate', your `.yaml` file should define only the classes present in your dataset. In your case, since 'license plate' is the only class, your `.yaml` should look like this:

```yaml
# data.yaml
train: ./train/images
val: ./valid/images
# Classes
names:
  0: license plate
```

There's no need to reserve indices for the COCO dataset classes unless you're combining your custom data with COCO. Happy training! 🚀

**Comment by ehsanadeli1** [comment]
@glenn-jocher Hi, thanks for your prompt response. Please modify your tutorial webpages. I spent hours correcting the use of 'train' and 'val' in my *.yaml file while following the tutorial (https://docs.ultralytics.com/datasets/detect/coco8/). Apparently, the correct way to call 'train' and 'val' in *.yaml is as follows:

#data.yaml
train: ./train
val: ./valid
#Classes
names:
  80: license plate

Someone mentioned it in the comments, and I was lucky to come across that comment, even though it wasn't at the top. Thanks again.

**Comment by ehsanadeli1** [comment]
Hi @glenn-jocher, I have another question. If I train a YOLOv8m model with images labeled 'license plate' only, will the retrained model still be able to detect the original 80 classes along with the new class 'license plate'? Or will it exclusively detect the new class 'license plate'? Thank you very much. Appreciate your time.

**Comment by glenn-jocher** [comment]
Hi @ehsanadeli1, if you train YOLOv8m exclusively on images labeled 'license plate', the resulting model will only be able to detect 'license plate' and not the original 80 classes. To detect both the original classes and 'license plate', you would need to include the original dataset with the 80 classes during training along with your new 'license plate' data. Happy training! 🚀

**Comment by ehsanadeli1** [comment]
Hi @glenn-jocher, Thank you for your assistance. I'm curious about how I can access the original dataset with the 80 classes. Is it publicly available, and if so, could you please provide the details? Thank you.

**Comment by ehsanadeli1** [comment]
Hi @glenn-jocher, Thank you for your prompt response. I'm curious about any performance comparisons between YOLOv8 trained on COCO, COCO128, and COCO8 datasets. Specifically, I'm wondering if models trained on COCO128 or COCO8 show a significant drop in performance compared to those trained on the full COCO dataset. Due to GPU limitations, I need to assess whether models trained on smaller datasets yield considerably lower performance or their performances are still high and reliable.

Additionally, could you provide an estimate of the number of epochs considered suitable for training these models on the three datasets? I am trying to gauge the computational cost, and any insights would be helpful.

As a side note, I plan to augment COCO128 with around 20 new pictures, introducing a total of 85 classes (80 main objects and 5 new objects). Your guidance on this matter is appreciated.

Thank you for your time.

**Comment by glenn-jocher** [comment]
Hi @ehsanadeli1, training on COCO128 or COCO8 will generally result in lower performance compared to the full COCO dataset due to the smaller variety and quantity of images. The extent of the performance drop depends on the complexity of the task and the diversity of the smaller dataset.

For epochs, it varies based on the dataset size. For COCO, hundreds of epochs are common, while for COCO128 or COCO8, fewer may be sufficient due to the smaller dataset size. However, the exact number should be determined empirically; start with our defaults and adjust based on your validation results.

When augmenting COCO128 with new classes, ensure your `.yaml` file reflects all 85 classes and that the new images are correctly labeled. Monitor the validation metrics to judge the model's performance on both the original and new classes. Happy training! 🚀

**Comment by ehsanadeli1** [comment]
Hi @glenn-jocher. I'm training a YOLOv8m on COCO datasets, focusing on three objects only: Person, Stop Sign, and Traffic Light. I have removed pictures from COCO images that don't contain any of these three classes. Therefore, training is now on only 66,807 images from COCO. My YAML file looks like this:

train: ../../../../datasets/coco_dl_objectsAdjusted/train
val: ../../../../datasets/coco_dl_objectsAdjusted/valid

# Classes
names:
  0: person
  1: traffic light
  2: stop sign

The issue is that after the first epoch of training, I see:

Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/100      1.68G     0.9903     0.9549      1.164         14        640: 100%|██████████| 4176/4176 [26:27<00:00,  2.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 704/704 [00:25<00:00, 28.01it/s]
                 
and after 44 epochs, I see:

    44/100      1.67G     0.9917     0.8431      1.177          1        640: 100%|██████████| 4176/4176 [11:25<00:00,  6.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 704/704 [00:21<00:00, 33.41it/s]


This shows, for example, that mAP50 increased from 0.582 to 0.706 only. So my question is, how high are the mAP50 or other metrics in models like YOLOv8m that you released? What values should I expect for these metrics, considering that for COCO datasets, you run with 80 objects, and I'm running with only 3 objects? (I know that closer to 1.0 is better) Thank you very much for your assistance. I appreciate it.

**Comment by glenn-jocher** [comment]
@ehsanadeli1 hi there! The mAP50 for models trained on the full COCO dataset with 80 classes is typically higher due to the larger variety of images and classes, which helps the model generalize better. For a model trained on only 3 classes, an mAP50 of 0.706 is quite good, but direct comparisons to the full COCO-trained models aren't entirely fair due to the differences in class variety and dataset size.

For your custom classes, ensure they are well-represented in your training data. The performance on your specific 3-class subset plus custom classes may differ from the full COCO dataset, but as long as it meets your application's requirements, you're on the right track. Keep an eye on the validation metrics to guide your training. Happy training! 🚀

**Comment by ehsanadeli1** [comment]
@glenn-jocher, thank you for your responses. I have a question regarding two cases:

Case 1: Utilizing all COCO datasets with 80 classes, including an additional class, and its corresponding images and labels that I add to the path.

Case 2: Utilizing a subset of COCO datasets related to the 3 classes mentioned in my previous message, along with one more class and its images and labels that I add to the path.

Clearly, training the first case is more expensive due to the higher number of images and labels. Now, for the following considerations:

a) What are your expectations for the performances of these two models? In your opinion, which model is likely to have better performance in detecting the 4 classes (Person, Stop Sign, Traffic Light, and the additional class), considering the high quality of images and labels for the new class? (I trained a YOLOv8m specifically for this new class, and it performed exceptionally well, but I aim for a single model to detect all 4 classes at once, rather than having two separate models—one for detecting 3 classes and another for the additional class.)

b) Do you have any other suggestions for achieving my goal?

**Comment by glenn-jocher** [comment]
@ehsanadeli1, for Case 1, the model is likely to generalize better across a wide range of classes, including the 4 you're interested in, due to the diverse training data. For Case 2, the model may perform slightly better on the 4 specific classes since it's more focused, but it might not generalize as well to new variations or similar classes.

For achieving your goal, I'd suggest starting with Case 2 for efficiency and if the performance is satisfactory for your application, stick with it. If not, consider Case 1 to leverage the broader generalization capabilities of the model. Remember to validate on a diverse set of images for the best assessment of performance. Happy training! 🚀

**Comment by shahbazahmed108** [comment]
@glenn-jocher  I trained yolov8 model on images with different classes and the .yaml for that data set was like this

train: C:\Users\name\Desktop\leica\train
val: C:\Users\name\Desktop\leica\val

nc: 13

names: ["ARG","WM","CW","FW-TM","CL","MW","KOCHIA","PL","PL-AS","PW","RTF","BROME","THISTLE"]

this training gave me the best.pt and that was running great on all 13 classes

then i have more images with a few classes as in the above training group and also some new classes. what I wanted to do is grab the already trained model as in the above dataset and train that further with a new dataset that shares some old classes and has some new ones. the number of classes for this dataset was 10 as PW, MW, PL, UB, CL, WM, WHEAT, HENBIT, NIGHTSHADE, BT. all these images are sitting in a new folder different from already trained model. that folder name is field_images - corrected-final

as you see some of the classes are new and some are old as in my first dataset. so to train the existing model with the already trained weight file name as best.pt, I created a new .yaml file with common classes with both datasets matching their indices and new classes were added at the end. so my new .yaml file for training looked like this

train: C:\Users\Shahbaz\Desktop\field_images - corrected-final\train
val: C:\Users\Shahbaz\Desktop\field_images - corrected-final\val

nc: 18

names: ["ARG","WM","CW","FW-TM","CL","MW","KOCHIA","PL","PL-AS","PW","RTF","BROME","THISTLE","UB","WHEAT","HENBIT","NIGHTSHADE","BT"]

remember the idea was to grab the best.pt file from the first trained model and train it further on new images which had some old classes and some new ones. for the second .yaml file, I arranged classes such that the old classes matched their place as old training and new classes were appended at the end. the problem is that when I ran the training on a new dataset hoping that the new best.pt will detect all old and new classes, it did not work and it is only detecting new classes from 18. the command that I used looked like this

yolo task=detect mode=train epochs=100 data=data_custom.yaml model="C:\Users\Shahbaz\Desktop\leica\best.pt" imgsz=640

The training process went fine but the detection is not happening as I hoped.
Please help me how to fix this

**Comment by ehsanadeli1** [comment]
@shahbazahmed108 Have you tried putting together all pictures of all classes you want to detect? Like you have in the first group 13 classes and the second group 10 classes. So your total objects would something between 13 and 23 classes in total. You should put all together and then make the .yaml file for all objects as well and run.  Please let us know if it gives you good results or not?

It's clear to me that when you retrain a new model with new classes the new model won't be able to detect previous classes. I asked this question in one of my messages above and @glenn-jocher confirmed that the new model is not able to learn the past classes.

**Comment by shahbazahmed108** [comment]
@ehsanadeli1 so I have to put the old images and new images together in one folder and run training on that? i thought I could just grab the already trained model and train it only with new images. No?

**Comment by shahbazahmed108** [comment]
say i trained my model on 100 images, next i want to further train it on 200 images, do i have to put together first 100 and second 200 and run training on those 300 images?

**Comment by ehsanadeli1** [comment]
@shahbazahmed108 Yea, this is my understanding that you should put all pictures together and the classes as well and train one single model. @glenn-jocher can confirm it though?

**Comment by shahbazahmed108** [comment]
@glenn-jocher @ehsanadeli1 but how was this guy able to do it in the issue https://github.com/ultralytics/ultralytics/issues/1631?

below is how he did it 
**********************
@glenn-jocher thank you I did it. I copied the best.pt and paste it to main folder with different name and I change the model name as you said and it did worked.

yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m.pt** imgsz=640 batch=11 patience=64

yolo task=detect mode=train epochs=128 data=data_custom.yaml model=**yolov8m_custom_train1.pt** imgsz=640 batch=11 patience=64
****************

i want to do the same thing, I don't know if his new images had the same classes or he had new.

**Comment by glenn-jocher** [comment]
@shahbazahmed108 The issue you're referring to seems to be about continuing training with a renamed model file, not about adding new classes to an existing model. When you add new classes and want to continue training, you need to retrain the model with both the old and new data together to ensure it learns to recognize all classes. If you only train on new images with new classes, the model will likely forget the old classes. This process is known as "incremental learning" or "fine-tuning," and it requires careful handling of the class indices and the dataset to maintain knowledge of all classes. 🧠
